### PR TITLE
Docs/user and contributing revamp

### PR DESCRIPTION
**Pull request by Dioprz**
### Description
This PR includes a complete update of our github documentation.

- **README improvements**: We now have safer installation instructions by removing the sudo recommendations, as well as a more inviting and friendly presentation for potential new contributors.

- **Structure refactor**: 
  - All the github documentation is now contained in a new `./docs/github/` directory.
  - Most of the previous documentation was splitted into related but independent topics, so it is easier to traverse and lighter to search and read.

- **Doctests documentation**: I removed all the deprecated documentation for Sage doctests, and create the new one based on Python doctests.

- **Doctstrings and HTML**: Was added a complete documentation on how to write a proper docstring for our library, as well as explanations on the particulars and conventions.

- **Additionals**:
  - I tried to mention and cover all the tech stacks, libraries and in-house solutions that we are using for the library, so it is easier to get onboard new contributors, but also to keep track of things for us, the existing ones. 
  - I included as much conventions as I was able to remember.

### Review process
- Start from the README, and just navigate for the new documentation.
- **User_guide.md**
  - I changed the documentation format, from Jupyter notebook to Markdown. Reasons include that not everyone has a jupyter session configured, our library can't be quickly installed for the moment (via pip), and we don't provide any setup intructions for such environment.
  - I'm not the most competent person to rewrite this particular documentation because I lack the knowledge. So please, review it and feel free to send your commits if you find something that can be improved.
- Take a look at the review comment with the pending discussion.

### Pre-approval checklist
- [x] The code builds clean without any errors or warnings
- [x] I've added/updated tests
- [x] I've added/updated documentation if needed

### CODE REVIEW

File: docs/github/code_conventions.md
Comment (Dioprz) [inline docs/github/code_conventions.md:1]: I took the freedom to add this here... Did you have any update on https://github.com/Crypto-TII/CryptographicEstimators/issues/163?

File: docs/github/code_conventions.md [hunk 1]
```diff
@@ -0,0 +1,34 @@
+# Code style
+
+- Line width: 120 characters.
+- Use absolute imports instead of relative imports.
```
Comment (Dioprz) [inline docs/github/code_conventions.md:4]: I took the freedom to add this here... Did you have any update on https://github.com/Crypto-TII/CryptographicEstimators/issues/163?

Comment (Javierverbel) [reply docs/github/code_conventions.md:4]: While running 
`sage test.py` as indicated in `add_new_estimator.md` I get a `ModuleNotFoundError:` error

File: docs/github/user_guide.md [hunk 2]
```diff
@@ -0,0 +1,986 @@
...
+```python
+Stern.memory_access = 2
+```
+
+```python
+Stern.time_complexity(r=2, p=3, l=4)
+```
+
+```
+36.85094946571111
```
Comment (Javierverbel) [inline docs/github/user_guide.md:430]: The result of thisone should be 29.89385629700021

File: docs/github/user_guide.md [hunk 3]
```diff
@@ -0,0 +1,986 @@
...
+{}
+```
+
+> **Hint:** The same method is applicable to the estimator object. It resets the
+> estimator as well as all associated algorithm objects
+
+```python
+SDE = SDEstimator(n=100, k=50, w=10)
+Stern = SDE.stern
+Stern.time_complexity()
```
Comment (Javierverbel) [inline docs/github/user_guide.md:556]: remove `Stern.time_complexity()`

File: docs/github/user_guide.md [hunk 4]
```diff
@@ -0,0 +1,986 @@
...
+ 'l': {'min': 0, 'max': 50},
+ 'p': {'min': 1, 'max': 3}}
+```
+
+```python
+Stern.optimal_parameters()
+```
+
+```
+{'r': 4, 'p': 1, 'l': 3}
```
Comment (Javierverbel) [inline docs/github/user_guide.md:636]: here it should be {'r': 4, 'p': 2, 'l': 9}

File: docs/github/user_guide.md [hunk 5]
```diff
@@ -0,0 +1,986 @@
...
+dictionary with the algorithm names, time and memory complexities, and the
+corresponding dictionary of optimal parameters.
+
+```python
+from cryptographic_estimators.MQEstimator import MQEstimator
+MQE = MQEstimator(n=15, m=17, q=3)
+MQE.estimate()
+```
+
+```
```
Comment (Javierverbel) [inline docs/github/user_guide.md:695]: Here the output should be 
```
{'BooleanSolveFXL': {'estimate': {'time': 29.858222427888588, 'memory': 12.901244032467376, 'parameters': {'k': 14, 'variant': 'las_vegas'}}, 'additional_information': {}}, 'Crossbred': {'estimate': {'time': 27.676784055214462, 'memory': 17.045780987724598, 'parameters': {'D': 3, 'd': 1, 'k': 6}}, 'additional_information': {}}, 'ExhaustiveSearch': {'estimate': {'time': 25.40490707466741, 'memory': 12.901244032467376, 'parameters': {}}, 'additional_information': {}}, 'F5': {'estimate': {'time': 48.247169726507984, 'memory': 31.484561900604888, 'parameters': {}}, 'additional_information': {}}, 'HybridF5': {'estimate': {'time': 27.605835266254303, 'memory': 12.901244032467376, 'parameters': {'k': 14}}, 'additional_information': {}}, 'Lokshtanov': {'estimate': {'time': 95.52681642861242, 'memory': 25.266724329355785, 'parameters': {'delta': 0.06666666666666667}}, 'additional_information': {}}}

```

File: docs/github/user_guide.md [hunk 6]
```diff
@@ -0,0 +1,986 @@
...
+In the next example we exclude the algorithms ExhaustiveSearch, F5, HybridF5,
+Lokshtanov.
+
+```python
+from cryptographic_estimators.MQEstimator import MQEstimator
+from cryptographic_estimators.MQEstimator.MQAlgorithms import ExhaustiveSearch, F5, HybridF5, Lokshtanov
+MQE = MQEstimator(n=15, m=17, q=3, excluded_algorithms = [ExhaustiveSearch, F5, HybridF5, Lokshtanov])
+MQE.estimate()
+```
+
```
Comment (Javierverbel) [inline docs/github/user_guide.md:734]: here the output should be 
```
{'BooleanSolveFXL': {'estimate': {'time': 29.858222427888588, 'memory': 12.901244032467376, 'parameters': {'k': 14, 'variant': 'las_vegas'}}, 'additional_information': {}}, 'Crossbred': {'estimate': {'time': 27.676784055214462, 'memory': 17.045780987724598, 'parameters': {'D': 3, 'd': 1, 'k': 6}}, 'additional_information': {}}}

```

File: docs/github/user_guide.md [hunk 7]
```diff
@@ -0,0 +1,986 @@
...
+and comparison.
+
+```python
+from cryptographic_estimators.MQEstimator import MQEstimator
+from cryptographic_estimators.MQEstimator.MQAlgorithms import BooleanSolveFXL, F5, Crossbred
+MQE = MQEstimator(n=15, m=17, q=3)
+MQE.table()
+```
+
+```
```
Comment (Javierverbel) [inline docs/github/user_guide.md:759]: Here the output should be
```
+------------------+---------------+
|                  |    estimate   |
+------------------+------+--------+
| algorithm        | time | memory |
+------------------+------+--------+
| BooleanSolveFXL  | 29.9 |   12.9 |
| Crossbred        | 27.7 |   17.0 |
| ExhaustiveSearch | 25.4 |   12.9 |
| F5               | 48.2 |   31.5 |
| HybridF5         | 27.6 |   12.9 |
| Lokshtanov       | 95.5 |   25.3 |
+------------------+------+--------+

```

File: docs/github/user_guide.md [hunk 8]
```diff
@@ -0,0 +1,986 @@
...
+
+Or again with some algorithm being excluded from consideration
+
+```python
+from cryptographic_estimators.MQEstimator.MQAlgorithms import ExhaustiveSearch, F5, HybridF5, Lokshtanov
+MQE = MQEstimator(n=15, m=17, q=3, excluded_algorithms = [ExhaustiveSearch, F5, HybridF5, Lokshtanov])
+MQE.table()
+```
+
+```
```
Comment (Javierverbel) [inline docs/github/user_guide.md:782]: Here the output should be

```
+-----------------+---------------+
|                 |    estimate   |
+-----------------+------+--------+
| algorithm       | time | memory |
+-----------------+------+--------+
| BooleanSolveFXL | 29.9 |   12.9 |
| Crossbred       | 27.7 |   17.0 |
+-----------------+------+--------+

```

File: docs/github/user_guide.md [hunk 9]
```diff
@@ -0,0 +1,986 @@
...
++---------------+--------+--------+-------------------------------------------------------+
+```
+
+4. `show_tilde_o_time` allows to show the time and memory estimates in the
+   TildeO mode.
+
+```python
+SDE.table(show_tilde_o_time=True)
+```
+
```
Comment (Javierverbel) [inline docs/github/user_guide.md:875]: Here the ouput should be

```
+---------------+------------------+---------------+
|               | tilde_o_estimate |    estimate   |
+---------------+-------+----------+------+--------+
| algorithm     |  time |   memory | time | memory |
+---------------+-------+----------+------+--------+
| BallCollision |  10.4 |      3.3 | 23.3 |   16.0 |
| BJMMdw        |    -- |       -- | 23.4 |   14.7 |
| BJMMpdw       |    -- |       -- | 23.3 |   14.3 |
| BJMM          |   9.0 |      6.8 | 22.8 |   15.0 |
| BJMM_plus     |    -- |       -- | 22.8 |   15.0 |
| BothMay       |   8.8 |      6.4 | 22.4 |   14.7 |
| Dumer         |  10.4 |      3.3 | 22.7 |   16.4 |
| MayOzerov     |   8.5 |      7.4 | 22.3 |   14.8 |
| Prange        |  10.8 |      0.0 | 28.3 |   12.7 |
| Stern         |  10.4 |      2.9 | 22.3 |   16.0 |
+---------------+-------+----------+------+--------+
```

File: docs/github/user_guide.md [hunk 10]
```diff
@@ -0,0 +1,986 @@
...
+```python
+MQE.crossbred.set_parameters({'d':1, 'D':3})
+```
+
+```python
+MQE.table(show_all_parameters=True)
+```
+
+```
```
Comment (Javierverbel) [inline docs/github/user_guide.md:949]: Here the output should be

``` 
+------------------+----------------------------------------------------+
|                  |                      estimate                      |
+------------------+-------+--------+-----------------------------------+
| algorithm        |  time | memory |             parameters            |
+------------------+-------+--------+-----------------------------------+
| BooleanSolveFXL  |  37.8 |   14.0 | {'k': 19, 'variant': 'las_vegas'} |
| Crossbred        |  35.7 |   18.0 |      {'d': 1, 'D': 3, 'k': 6}     |
| ExhaustiveSearch |  33.5 |   14.0 |                 {}                |
| F5               |  60.0 |   39.7 |                 {}                |
| HybridF5         |  35.8 |   14.0 |             {'k': 19}             |
| Lokshtanov       | 106.6 |   33.2 |          {'delta': 0.05}          |
+------------------+-------+--------+-----------------------------------+

```

File: docs/github/user_guide.md [hunk 11]
```diff
@@ -0,0 +1,986 @@
...
+Crossbred estimator for the MQ problem with 20 variables and 20 polynomials
+```
+
+```python
+F.time_complexity()
+```
+
+```
```
Comment (Javierverbel) [inline docs/github/user_guide.md:984]: output should be `33.475373791757825`

**Review by Javierverbel — APPROVED** [review]
After minor comments are addressed it can be marged @Diego

**Review by Javierverbel — CHANGES_REQUESTED** [review]
Some hyperlinks in the Readme don't work.
For instance:
```
Look for contributing documentation in its [own section](https://github.com/Crypto-TII/CryptographicEstimators/pull/212/files/7b5725f5f387cc77230de7870f2eedc5b0c49f80#Contributing-%F0%9F%A4%9D).
```

**Review by Javierverbel — APPROVED** [review]

**Pull request merged**